File: convex-backend/convex/queue.ts
import { mutation, query, internalMutation } from "./_generated/server";
import { internal } from "./_generated/api";
import { v } from "convex/values";

const CLAIM_TIMEOUT_MS = 10 * 60 * 1000; // 10 minutes

/**
 * Clean up stale claims older than CLAIM_TIMEOUT_MS
 */
export const cleanupStaleClaims = internalMutation({
  args: {},
  handler: async (ctx) => {
    const now = Date.now();
    const cutoff = now - CLAIM_TIMEOUT_MS;

    // Clean stale SQ claims
    const staleSqClaims = await ctx.db
      .query("sq_claims")
      .withIndex("by_claimed_at")
      .filter((q) => q.lt(q.field("claimedAt"), cutoff))
      .collect();

    for (const claim of staleSqClaims) {
      await ctx.db.delete(claim._id);
    }

    // Clean stale refund reservations
    const staleRefunds = await ctx.db
      .query("refund_reservations")
      .withIndex("by_reserved_at")
      .filter((q) => q.lt(q.field("reservedAt"), cutoff))
      .collect();

    for (const reservation of staleRefunds) {
      await ctx.db.delete(reservation._id);
    }

    return {
      deletedSqClaims: staleSqClaims.length,
      deletedRefundReservations: staleRefunds.length,
    };
  },
});

/**
 * EMERGENCY: Clean up ALL claims and reservations (use when system is stuck)
 */
export const forceCleanupAll = mutation({
  args: {},
  handler: async (ctx) => {
    // Delete ALL SQ claims
    const allSqClaims = await ctx.db.query("sq_claims").collect();
    for (const claim of allSqClaims) {
      await ctx.db.delete(claim._id);
    }

    // Delete ALL refund reservations
    const allRefunds = await ctx.db.query("refund_reservations").collect();
    for (const reservation of allRefunds) {
      await ctx.db.delete(reservation._id);
    }

    return {
      deletedSqClaims: allSqClaims.length,
      deletedRefundReservations: allRefunds.length,
      message: "All claims and reservations cleared",
    };
  },
});

/**
 * Internal mutation wrapper for tryClaimSQ (for HTTP actions)
 */
export const tryClaimSQInternal = internalMutation({
  args: {
    botId: v.string(),
    sqNumber: v.string(),
  },
  handler: async (ctx, args) => {
    // Clean up stale claims first
    await ctx.runMutation(internal.queue.cleanupStaleClaims, {});

    // Check if already claimed (get ALL claims for this SQ, not just first)
    const existingClaims = await ctx.db
      .query("sq_claims")
      .withIndex("by_sq_number", (q) => q.eq("sqNumber", args.sqNumber))
      .collect();

    // Check for any active claims (CLAIMING status)
    const activeClaim = existingClaims.find(c => c.status === "CLAIMING");
    if (activeClaim) {
      return {
        success: false,
        message: `SQ ${args.sqNumber} already claimed by ${activeClaim.botId}`,
        claimedBy: activeClaim.botId,
      };
    }

    // Check for recent COMPLETED claims (within last 60 seconds)
    const now = Date.now();
    const recentCompleted = existingClaims.find(c =>
      c.status === "COMPLETED" &&
      c.completedAt &&
      (now - c.completedAt) < 60000
    );

    if (recentCompleted) {
      return {
        success: false,
        message: `SQ ${args.sqNumber} was recently completed by ${recentCompleted.botId}`,
        claimedBy: recentCompleted.botId,
      };
    }

    // Claim it
    await ctx.db.insert("sq_claims", {
      sqNumber: args.sqNumber,
      botId: args.botId,
      status: "CLAIMING",
      claimedAt: Date.now(),
    });

    return {
      success: true,
      message: `Successfully claimed SQ ${args.sqNumber}`,
      sqNumber: args.sqNumber,
      botId: args.botId,
    };
  },
});

/**
 * Try to claim an SQ for a bot (client-facing mutation)
 */
export const tryClaimSQ = mutation({
  args: {
    botId: v.string(),
    sqNumber: v.string(),
  },
  handler: async (ctx, args) => {
    // Clean up stale claims first
    await ctx.runMutation(internal.queue.cleanupStaleClaims, {});

    // Check if already claimed (get ALL claims for this SQ, not just first)
    const existingClaims = await ctx.db
      .query("sq_claims")
      .withIndex("by_sq_number", (q) => q.eq("sqNumber", args.sqNumber))
      .collect();

    // Check for any active claims (CLAIMING status)
    const activeClaim = existingClaims.find(c => c.status === "CLAIMING");
    if (activeClaim) {
      return {
        success: false,
        message: `SQ ${args.sqNumber} already claimed by ${activeClaim.botId}`,
        claimedBy: activeClaim.botId,
      };
    }

    // Check for recent COMPLETED claims (within last 60 seconds)
    // This prevents race condition where BOT1 completes and BOT2 tries to claim immediately
    const now = Date.now();
    const recentCompleted = existingClaims.find(c =>
      c.status === "COMPLETED" &&
      c.completedAt &&
      (now - c.completedAt) < 60000
    );

    if (recentCompleted) {
      return {
        success: false,
        message: `SQ ${args.sqNumber} was recently completed by ${recentCompleted.botId}`,
        claimedBy: recentCompleted.botId,
      };
    }

    // Claim it
    await ctx.db.insert("sq_claims", {
      sqNumber: args.sqNumber,
      botId: args.botId,
      status: "CLAIMING",
      claimedAt: Date.now(),
    });

    return {
      success: true,
      message: `Successfully claimed SQ ${args.sqNumber}`,
      sqNumber: args.sqNumber,
      botId: args.botId,
    };
  },
});

/**
 * Internal mutation wrapper for releaseSQ (for HTTP actions)
 */
export const releaseSQInternal = internalMutation({
  args: {
    botId: v.string(),
    sqNumber: v.string(),
  },
  handler: async (ctx, args) => {
    const claim = await ctx.db
      .query("sq_claims")
      .withIndex("by_sq_number", (q) => q.eq("sqNumber", args.sqNumber))
      .first();

    if (!claim) {
      return {
        success: false,
        message: `No claim found for SQ ${args.sqNumber}`,
      };
    }

    if (claim.botId !== args.botId) {
      return {
        success: false,
        message: `SQ ${args.sqNumber} claimed by ${claim.botId}, not ${args.botId}`,
      };
    }

    // Mark as completed
    await ctx.db.patch(claim._id, {
      status: "COMPLETED",
      completedAt: Date.now(),
    });

    return {
      success: true,
      message: `Released SQ ${args.sqNumber}`,
    };
  },
});

/**
 * Release an SQ claim (client-facing mutation)
 */
export const releaseSQ = mutation({
  args: {
    botId: v.string(),
    sqNumber: v.string(),
  },
  handler: async (ctx, args) => {
    const claim = await ctx.db
      .query("sq_claims")
      .withIndex("by_sq_number", (q) => q.eq("sqNumber", args.sqNumber))
      .first();

    if (!claim) {
      return {
        success: false,
        message: `No claim found for SQ ${args.sqNumber}`,
      };
    }

    if (claim.botId !== args.botId) {
      return {
        success: false,
        message: `SQ ${args.sqNumber} claimed by ${claim.botId}, not ${args.botId}`,
      };
    }

    // Mark as completed
    await ctx.db.patch(claim._id, {
      status: "COMPLETED",
      completedAt: Date.now(),
    });

    return {
      success: true,
      message: `Released SQ ${args.sqNumber}`,
    };
  },
});

/**
 * Reserve rows in Refund Log
 */
export const reserveRefundLogWrite = mutation({
  args: {
    botId: v.string(),
    sqNumber: v.string(),
    rowCount: v.number(),
    currentLastRow: v.number(),
  },
  handler: async (ctx, args) => {
    // Clean up stale claims first
    await ctx.runMutation(internal.queue.cleanupStaleClaims, {});

    // Calculate next available row
    let nextRow = args.currentLastRow + 1;

    // Find highest reserved row from active reservations
    const activeReservations = await ctx.db
      .query("refund_reservations")
      .withIndex("by_status", (q) => q.eq("status", "WRITING"))
      .collect();

    for (const reservation of activeReservations) {
      const endRow = reservation.startRow + reservation.rowCount;
      if (endRow > nextRow) {
        nextRow = endRow;
      }
    }

    // Create reservation
    await ctx.db.insert("refund_reservations", {
      sqNumber: args.sqNumber,
      botId: args.botId,
      startRow: nextRow,
      rowCount: args.rowCount,
      status: "WRITING",
      reservedAt: Date.now(),
    });

    return {
      success: true,
      startRow: nextRow,
      rowCount: args.rowCount,
      sqNumber: args.sqNumber,
      botId: args.botId,
    };
  },
});

/**
 * Release Refund Log reservation
 */
export const releaseRefundLogWrite = mutation({
  args: {
    botId: v.string(),
    sqNumber: v.string(),
  },
  handler: async (ctx, args) => {
    const reservation = await ctx.db
      .query("refund_reservations")
      .withIndex("by_sq_number", (q) => q.eq("sqNumber", args.sqNumber))
      .first();

    if (!reservation) {
      return {
        success: false,
        message: `No reservation found for SQ ${args.sqNumber}`,
      };
    }

    if (reservation.botId !== args.botId) {
      return {
        success: false,
        message: `Reservation for SQ ${args.sqNumber} owned by ${reservation.botId}, not ${args.botId}`,
      };
    }

    // Mark as completed
    await ctx.db.patch(reservation._id, {
      status: "COMPLETED",
      completedAt: Date.now(),
    });

    return {
      success: true,
      message: `Released Refund Log reservation for SQ ${args.sqNumber}`,
    };
  },
});

/**
 * Get list of currently claimed SQ numbers (for preventing duplicate attempts)
 * Returns array of SQ numbers that are currently CLAIMING (not yet COMPLETED)
 */
export const getClaimedSQs = query({
  args: {},
  handler: async (ctx) => {
    // Get all active (CLAIMING status) SQ claims
    const activeClaims = await ctx.db
      .query("sq_claims")
      .withIndex("by_status", (q) => q.eq("status", "CLAIMING"))
      .collect();

    // Return just the SQ numbers
    return activeClaims.map(claim => claim.sqNumber);
  },
});

/**
 * Get queue status (for debugging)
 */
export const getQueueStatus = query({
  args: {},
  handler: async (ctx) => {
    const sqClaims = await ctx.db.query("sq_claims").collect();
    const refundReservations = await ctx.db.query("refund_reservations").collect();

    return {
      success: true,
      sqClaims: sqClaims.map((c) => ({
        sqNumber: c.sqNumber,
        botId: c.botId,
        status: c.status,
        claimedAt: new Date(c.claimedAt).toISOString(),
        completedAt: c.completedAt ? new Date(c.completedAt).toISOString() : undefined,
      })),
      refundReservations: refundReservations.map((r) => ({
        sqNumber: r.sqNumber,
        botId: r.botId,
        startRow: r.startRow,
        rowCount: r.rowCount,
        status: r.status,
        reservedAt: new Date(r.reservedAt).toISOString(),
        completedAt: r.completedAt ? new Date(r.completedAt).toISOString() : undefined,
      })),
      timestamp: new Date().toISOString(),
    };
  },
});

File: pdf-parser-server/api/queue.py
"""
Vercel Queue Management API for Helper Doc Coordination

Uses Vercel KV (Redis) for persistent, atomic queue coordination across
all function instances and concurrent requests.

Endpoints:
- POST /api/queue - Manage SQ claims and Refund Log reservations
- GET /api/queue - Health check and queue status
"""

from http.server import BaseHTTPRequestHandler
import json
from datetime import datetime, timedelta
import os
import requests

# Vercel KV REST API credentials (automatically set by Vercel)
KV_REST_API_URL = os.environ.get('KV_REST_API_URL', '')
KV_REST_API_TOKEN = os.environ.get('KV_REST_API_TOKEN', '')

CLAIM_TIMEOUT_SECONDS = 600  # 10 minutes

class VercelKV:
    """Vercel KV REST API client"""

    def __init__(self, url, token):
        self.url = url.rstrip('/')
        self.headers = {
            'Authorization': f'Bearer {token}',
            'Content-Type': 'application/json'
        }

    def get(self, key):
        """Get value for key"""
        try:
            response = requests.get(
                f'{self.url}/get/{key}',
                headers=self.headers,
                timeout=5
            )
            if response.status_code == 200:
                result = response.json()
                return result.get('result')
            return None
        except Exception as e:
            print(f'KV GET error: {e}')
            return None

    def set(self, key, value, ex=None):
        """Set key to value with optional expiration in seconds"""
        try:
            data = [key, value]
            if ex:
                data.extend(['EX', ex])

            response = requests.post(
                f'{self.url}/set',
                headers=self.headers,
                json=data,
                timeout=5
            )
            return response.status_code == 200
        except Exception as e:
            print(f'KV SET error: {e}')
            return False

    def delete(self, key):
        """Delete key"""
        try:
            response = requests.post(
                f'{self.url}/del',
                headers=self.headers,
                json=[key],
                timeout=5
            )
            return response.status_code == 200
        except Exception as e:
            print(f'KV DEL error: {e}')
            return False

    def keys(self, pattern='*'):
        """Get all keys matching pattern"""
        try:
            response = requests.get(
                f'{self.url}/keys/{pattern}',
                headers=self.headers,
                timeout=5
            )
            if response.status_code == 200:
                result = response.json()
                return result.get('result', [])
            return []
        except Exception as e:
            print(f'KV KEYS error: {e}')
            return []

# Initialize KV client
kv = VercelKV(KV_REST_API_URL, KV_REST_API_TOKEN) if KV_REST_API_URL and KV_REST_API_TOKEN else None

def clean_stale_claims():
    """Remove claims older than CLAIM_TIMEOUT_SECONDS"""
    if not kv:
        return

    now = datetime.now()

    # Clean SQ claims
    sq_keys = kv.keys('sq:*')
    for key in sq_keys:
        data = kv.get(key)
        if data:
            try:
                claim = json.loads(data)
                claim_time = datetime.fromisoformat(claim['timestamp'])
                if (now - claim_time).total_seconds() > CLAIM_TIMEOUT_SECONDS:
                    kv.delete(key)
            except:
                pass

    # Clean Refund Log reservations
    refund_keys = kv.keys('refund:*')
    for key in refund_keys:
        data = kv.get(key)
        if data:
            try:
                reservation = json.loads(data)
                reservation_time = datetime.fromisoformat(reservation['timestamp'])
                if (now - reservation_time).total_seconds() > CLAIM_TIMEOUT_SECONDS:
                    kv.delete(key)
            except:
                pass

def try_claim_sq(bot_id, sq_number):
    """Try to claim an SQ for a bot"""
    if not kv:
        return {
            'success': False,
            'error': 'KV not configured. Set up Vercel KV in dashboard.'
        }

    clean_stale_claims()

    key = f'sq:{sq_number}'

    # Check if already claimed
    existing = kv.get(key)
    if existing:
        try:
            claim = json.loads(existing)
            if claim['status'] == 'CLAIMING':
                return {
                    'success': False,
                    'message': f"SQ {sq_number} already claimed by {claim['botId']}",
                    'claimedBy': claim['botId']
                }
        except:
            pass

    # Claim it
    claim = {
        'botId': bot_id,
        'timestamp': datetime.now().isoformat(),
        'status': 'CLAIMING'
    }

    # Set with expiration (auto-cleanup after timeout)
    if kv.set(key, json.dumps(claim), ex=CLAIM_TIMEOUT_SECONDS):
        return {
            'success': True,
            'message': f"Successfully claimed SQ {sq_number}",
            'sqNumber': sq_number,
            'botId': bot_id
        }
    else:
        return {
            'success': False,
            'error': 'Failed to write to KV'
        }

def release_sq(bot_id, sq_number):
    """Release an SQ claim"""
    if not kv:
        return {'success': False, 'error': 'KV not configured'}

    key = f'sq:{sq_number}'

    # Check ownership
    existing = kv.get(key)
    if not existing:
        return {
            'success': False,
            'message': f"No claim found for SQ {sq_number}"
        }

    try:
        claim = json.loads(existing)
        if claim['botId'] != bot_id:
            return {
                'success': False,
                'message': f"SQ {sq_number} claimed by {claim['botId']}, not {bot_id}"
            }
    except:
        pass

    # Mark as completed (keep for debugging)
    claim['status'] = 'COMPLETED'
    claim['completedAt'] = datetime.now().isoformat()
    kv.set(key, json.dumps(claim), ex=CLAIM_TIMEOUT_SECONDS)

    return {
        'success': True,
        'message': f"Released SQ {sq_number}"
    }

def reserve_refund_log_write(bot_id, sq_number, row_count, current_last_row=1):
    """
    Reserve rows in Refund Log - returns starting row number

    Args:
        bot_id: Bot identifier (e.g., 'BOT1')
        sq_number: SQ number being processed
        row_count: Number of rows needed
        current_last_row: Last used row from Refund Log sheet

    Returns:
        Starting row number for this reservation
    """
    if not kv:
        return {'success': False, 'error': 'KV not configured'}

    clean_stale_claims()

    # Calculate next available row
    next_row = current_last_row + 1

    # Find highest reserved row from active reservations
    refund_keys = kv.keys('refund:*')
    for key in refund_keys:
        data = kv.get(key)
        if data:
            try:
                reservation = json.loads(data)
                if reservation.get('status') != 'COMPLETED':
                    end_row = reservation['startRow'] + reservation['rowCount']
                    if end_row > next_row:
                        next_row = end_row
            except:
                pass

    # Create reservation
    reservation = {
        'botId': bot_id,
        'startRow': next_row,
        'rowCount': row_count,
        'status': 'WRITING',
        'timestamp': datetime.now().isoformat()
    }

    key = f'refund:{sq_number}'
    if kv.set(key, json.dumps(reservation), ex=CLAIM_TIMEOUT_SECONDS):
        return {
            'success': True,
            'startRow': next_row,
            'rowCount': row_count,
            'sqNumber': sq_number,
            'botId': bot_id
        }
    else:
        return {
            'success': False,
            'error': 'Failed to write reservation to KV'
        }

def release_refund_log_write(bot_id, sq_number):
    """Mark Refund Log reservation as completed"""
    if not kv:
        return {'success': False, 'error': 'KV not configured'}

    key = f'refund:{sq_number}'

    # Check ownership
    existing = kv.get(key)
    if not existing:
        return {
            'success': False,
            'message': f"No reservation found for SQ {sq_number}"
        }

    try:
        reservation = json.loads(existing)
        if reservation['botId'] != bot_id:
            return {
                'success': False,
                'message': f"Reservation for SQ {sq_number} owned by {reservation['botId']}, not {bot_id}"
            }
    except:
        pass

    # Mark as completed
    reservation['status'] = 'COMPLETED'
    reservation['completedAt'] = datetime.now().isoformat()
    kv.set(key, json.dumps(reservation), ex=CLAIM_TIMEOUT_SECONDS)

    return {
        'success': True,
        'message': f"Released Refund Log reservation for SQ {sq_number}"
    }

def get_queue_status():
    """Get current queue status (for debugging)"""
    if not kv:
        return {
            'success': False,
            'error': 'KV not configured. Visit Vercel dashboard to set up KV storage.'
        }

    clean_stale_claims()

    # Get all SQ claims
    sq_claims = {}
    sq_keys = kv.keys('sq:*')
    for key in sq_keys:
        data = kv.get(key)
        if data:
            try:
                sq_number = key.replace('sq:', '')
                sq_claims[sq_number] = json.loads(data)
            except:
                pass

    # Get all Refund Log reservations
    refund_reservations = {}
    refund_keys = kv.keys('refund:*')
    for key in refund_keys:
        data = kv.get(key)
        if data:
            try:
                sq_number = key.replace('refund:', '')
                refund_reservations[sq_number] = json.loads(data)
            except:
                pass

    return {
        'success': True,
        'sqClaims': sq_claims,
        'refundReservations': refund_reservations,
        'timestamp': datetime.now().isoformat(),
        'kvConfigured': True
    }

class handler(BaseHTTPRequestHandler):
    def do_GET(self):
        """Health check and status endpoint"""
        self.send_response(200)
        self.send_header('Content-type', 'application/json')
        self.send_header('Access-Control-Allow-Origin', '*')
        self.end_headers()

        response = get_queue_status()
        self.wfile.write(json.dumps(response).encode())
        return

    def do_POST(self):
        """Handle queue management requests"""
        try:
            # Read request body
            content_length = int(self.headers.get('Content-Length', 0))
            body = self.rfile.read(content_length).decode('utf-8')
            data = json.loads(body)

            action = data.get('action')
            bot_id = data.get('botId')
            sq_number = data.get('sqNumber')

            if not action:
                self.send_response(400)
                self.send_header('Content-type', 'application/json')
                self.end_headers()
                self.wfile.write(json.dumps({
                    'success': False,
                    'error': 'Missing action parameter'
                }).encode())
                return

            # Route to appropriate handler
            if action == 'tryClaimSQ':
                result = try_claim_sq(bot_id, sq_number)

            elif action == 'releaseSQ':
                result = release_sq(bot_id, sq_number)

            elif action == 'reserveRefundLogWrite':
                row_count = data.get('rowCount', 1)
                current_last_row = data.get('currentLastRow', 1)
                result = reserve_refund_log_write(bot_id, sq_number, row_count, current_last_row)

            elif action == 'releaseRefundLogWrite':
                result = release_refund_log_write(bot_id, sq_number)

            elif action == 'getStatus':
                result = get_queue_status()

            else:
                result = {
                    'success': False,
                    'error': f'Unknown action: {action}'
                }

            # Send response
            self.send_response(200)
            self.send_header('Content-type', 'application/json')
            self.send_header('Access-Control-Allow-Origin', '*')
            self.end_headers()
            self.wfile.write(json.dumps(result).encode())

        except Exception as e:
            self.send_response(500)
            self.send_header('Content-type', 'application/json')
            self.end_headers()
            self.wfile.write(json.dumps({
                'success': False,
                'error': str(e)
            }).encode())

    def do_OPTIONS(self):
        """Handle CORS preflight"""
        self.send_response(200)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET, POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')
        self.end_headers()
